src/core/cgroup.c

```diff
@@ -1021,19 +1039,26 @@ static void cgroup_context_apply(
         }
 
         if ((apply_mask & CGROUP_MASK_DEVICES) && !is_root) {
+                _cleanup_(bpf_program_unrefp) BPFProgram *prog = NULL;
                 CGroupDeviceAllow *a;
 
-                /* Changing the devices list of a populated cgroup
-                 * might result in EINVAL, hence ignore EINVAL
-                 * here. */
+                if (cg_all_unified() > 0) {
+                        r = cgroup_init_device_bpf(&prog, c->device_policy, c->device_allow);
+                        if (r < 0)
+                                log_unit_full(u, LOG_WARNING, r, "Failed to initialize device control bpf program: %m");
```

Comment:
cgroup_apply_device_bpf() is the function, which will fail in case that the kernel doesn't support the bpf-based device controller. cgroup_init_device_bpf() will work in any case. I'll add a printed once notice there. Probably, long-term it's better to extend bpf_firewall_supported() to get all info about supported bpf functionality, and use it for any kinds of bpf programs. Thanks!

Reply:
BTW, why don't you use `log_unit_warning_errno()` here?

Reply:
yes, @yuwata is right, please use log_unit_warning_errno() unless you have a reson to prefer the full version...